sdk/linux_kernel_drivers/xdma/cdev_xvc.c

```diff
@@ -190,44 +197,46 @@ static long xvc_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 		/* Shift data out and copy to output buffer */
 		rv = xvc_shift_bits(iobase, tms_store, tdi_store, &tdo_store);
 		if (rv < 0)
-			goto cleanup;
+			break;
 
 		memcpy(tdo_buf + bytes, &tdo_store, shift_bytes);
 	}
 
+	if (rv < 0)
+		goto unlock;
+
 	/* if testing bar access swap tdi and tdo bufferes to "loopback" */
 	if (opcode == 0x2) {
-		char *tmp = tdo_buf;
+		unsigned char *tmp = tdo_buf;
 
 		tdo_buf = tdi_buf;
 		tdi_buf = tmp;
 	}
 
-	rv = copy_to_user((void *)xvc_obj.tdo_buf, tdo_buf, total_bytes);
-	if (rv) {
+	rv = copy_to_user(xvc_obj.tdo_buf, (const void *)tdo_buf, total_bytes);
+	if (rv)
 		pr_info("copy back tdo_buf failed: %d/%u.\n", rv, total_bytes);
-		rv = -EFAULT;
-		goto cleanup;
-	}
-
-cleanup:
-	if (buffer)
-		kfree(buffer);
 
+unlock:
+#if KERNEL_VERSION(5, 1, 0) >= LINUX_VERSION_CODE
 	mmiowb();
```

Comment:
@karenx-xilinx  priority to kernel 5.1,  once would use dma_wmb()    (or just wmb) ,  that what other device drivers used before intro of mmiowb

Reply:
Thanks. Will change it.

Reply:
Change is in place, resolving conversation.